pydatastructs/linear_data_structures/algorithms.py

```diff
@@ -1369,3 +1372,197 @@ def bubble_sort(array, **kwargs):
+        The array which is to be sorted.
+    elem: same as that of the array
+    start: int
+        The starting index of the portion
+        which is to be searched.
+        Optional, by default 0
+    end: int
+        The ending index of the portion which
+        is to be searched.
+        Optional, by default the index
+        of the last position filled.
+    backend: pydatastructs.Backend
+        The backend to be used.
+        Optional, by default, the best available
+        backend is used.
+
+    Returns
+    =======
+
+    output: int
+        The index of elem if found.
+        If not found, returns -1.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import OneDimensionalArray, linear_search
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> out = linear_search(arr, 2)
+    >>> print(out)
+    1
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Linear_search
+    """
+    raise_if_backend_is_not_python(
+            linear_search, kwargs.get('backend', Backend.PYTHON))
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    for i in range(start, end+1):
+        if array[i] == elem:
+            return i
+    return -1
+
+def binary_search(array, elem, **kwargs):
+    """
+    Implements binary search algorithm ON A SORTED ARRAY.
+
+    Parameters
+    ==========
+
+    array: OneDimensionalArray
+        The array which is to be sorted.
+    elem: same as that of the array
+    start: int
+        The starting index of the portion
+        which is to be searched.
+        Optional, by default 0
+    end: int
+        The ending index of the portion which
+        is to be searched.
+        Optional, by default the index
+        of the last position filled.
+    comp: lambda/function
+        The comparator which is to be used
+        for sorting. If the function returns
+        False then only swapping is performed.
+        Optional, by default, less than or
+        equal to is used for comparing two
+        values.
+    backend: pydatastructs.Backend
+        The backend to be used.
+        Optional, by default, the best available
+        backend is used.
+
+    Returns
+    =======
+
+    output: int
+        The index of elem if found.
+        If not found, returns -1.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import OneDimensionalArray, binary_search
+    >>> arr = OneDimensionalArray(int,[1, 2, 3, 5, 10, 12])
+    >>> out = binary_search(arr, 5)
+    >>> print(out)
+    3
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Binary_search_algorithm
+    """
+    raise_if_backend_is_not_python(
+            binary_search, kwargs.get('backend', Backend.PYTHON))
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+    Left = start
+    Right = end
+    while Left <= Right:
+        Middle = int((Left+Right)/2)
```

Comment:
What about, `Left//2 + Right//2`?

Reply:
I think (Left + Right)//2 is fine but not too sure about the sum of floors, for example, when left = 3, right = 5(when both odd), it will give 3 as middle

Reply:
Well the loss won't be much. There are four cases,

1. Left - Odd, Right - Odd - Middle will be lesser (by 1) in Left//2 + Right//2 as compared to current approach.
2. Left - Odd, Right - Even or Right - Odd, Left - Even - Middle will be exactly the same as in current approach.
3. Left - Even, Right - Even - Middle will again be exactly the same as in current approach.

The reason is (Left + Right) might get too large for implicit arrays leading to slow down in performance (Python uses big integers for handling larger than 64 bit integers). So, basically if you can just add 1 for Left - Odd, Right - Odd in Left//2 + Right//2 and do only Left//2 + Right//2 for other cases then it should be fine. In fact,

`Middle = Left//2 + Right//2 + Left % 2 * Right % 2` should work. Worth trying.

Reply:
Yes, this seems much better, I shall do this